File: 3.Hard/4. Median of Two Sorted Arrays.cpp
#include<iostream>
#include<vector>
using namespace std;
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        vector<int> arr;
        int m = nums1.size();
        int n = nums2.size();
        int id1 = 0 ;
        int id2 = 0;
        
        while(id1 < m && id2 < n)
        {
        	if(nums1[id1] < nums2[id2]) arr.push_back(nums1[id1++]);
        	else arr.push_back(nums2[id2++]);
		}
        while(id1 < m) arr.push_back(nums1[id1++]);
        while(id2 < n) arr.push_back(nums2[id2++]);
        for(int i = 0;i < arr.size();i ++) cout<<arr[i]<<" ";
        
        if(arr.size() % 2 == 1) return (double)arr[arr.size()/2];
        else return (double(arr[arr.size()/2] + double(arr[arr.size()/2-1]))) /2; 
    }
};

int main()
{
	vector<int> a1 = {1,2,3,4};
	vector<int> a2 = {2,3,4,5,6};
	Solution().findMedianSortedArrays(a1,a2);
}

File: 3.Hard/4. Median of Two Sorted Arrays.py
'''Solution: Similar to mergeing two sorted arr and find the number of index (l1 + l2)//2 to find the median
Complexity : O(n)'''
class Solution:
    def findMedianSortedArrays(self, nums1: list[int], nums2: list[int]) -> float:
        l1 = len(nums1)
        l2 = len(nums2)
        idmax = (l1+l2) // 2
        
        id = 0
        id1 = 0
        id2 = 0  
        pre = now = min(nums1[0],nums2[0])

        for i in range(min(l1,l2)):
            if nums1[id1] < nums2[id2]:
                pre = now
                now = nums1[id1]
                id1 += 1
            else :
                pre = now 
                now = nums2[id2]
                id2 += 1
            id += 1
        while id1 < l1 and id <= idmax:
            pre = now 
            now = nums1[id1]
            id1 += 1
            id += 1
        while id2 < l2 and id <= idmax:
            pre = now 
            now = nums2[id2]
            id2 += 1
            id += 1
        
        if idmax & 1 :
            return float (now)
        return (now + pre)/2

                    

print(Solution().findMedianSortedArrays([1,3], [2]))